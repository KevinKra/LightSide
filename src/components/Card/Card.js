import React from "react";
import "./Card.scss";
import PropTypes from "prop-types";

export default function Card(props) {
  const {
    name,
    gender,
    hair_color,
    eye_color,
    birth_year,
    mass,
    height,
    climate,
    terrain,
    population,
    rotation_period,
    orbital_period,
    diameter,
    model,
    vehicle_class,
    manufacturer,
    cost_in_credits,
    max_atmosphering_speed,
    crew,
    passengers,
    cargo_capacity
  } = props.element;

  if (props.format === "person") {
    return (
      <article className="Card">
        <section className="card-primary">
          <h1>{name}</h1>
          <p>gender: {gender}</p>
          <p>hair: {hair_color}</p>
          <p>eye color: {eye_color}</p>
          <p>birth year: {birth_year}</p>
          <p>mass: {mass}</p>
          <p>height: {height}</p>
        </section>
        <div className="card-bg person-bg" />
        {!props.element.favorited ? (
          <button
            className="favorite"
            onClick={() => props.addToFavorites(name)}
          >
            Favorite
          </button>
        ) : (
          <button
            className="unfavorite"
            onClick={() => props.removeFromFavorites(name)}
          >
            ✓
          </button>
        )}
      </article>
    );
  }
  if (props.format === "planet") {
    return (
      <article className="Card">
        <section className="card-primary">
          <h1>{name}</h1>
          <p>climate: {climate}</p>
          <p>terrain: {terrain}</p>
          <p>population: {population}</p>
          <p>day length: {rotation_period} hours</p>
          <p>year length: {orbital_period} days</p>
          <p>diameter: {diameter}</p>
        </section>
        <div className="card-bg planet-bg" />
        {!props.element.favorited ? (
          <button
            className="favorite"
            onClick={() => props.addToFavorites(name)}
          >
            Favorite
          </button>
        ) : (
          <button
            className="unfavorite"
            onClick={() => props.removeFromFavorites(name)}
          >
            ✓
          </button>
        )}
      </article>
    );
  }
  if (props.format === "vehicle") {
    return (
      <article className="Card">
        <section className="card-primary">
          <h1>{name}</h1>
          <p>model: {model}</p>
          <p>class: {vehicle_class}</p>
          <p>manufacturer: {manufacturer}</p>
          <p>cost: {cost_in_credits}</p>
          <p>maximum speed: {max_atmosphering_speed}</p>
          <p>crew capacity: {crew}</p>
          <p>passenger capacity: {passengers}</p>
          <p>cargo capacity: {cargo_capacity}</p>
        </section>
        <div className="card-bg vehicle-bg" />
        {!props.element.favorited ? (
          <button
            className="favorite"
            onClick={() => props.addToFavorites(name)}
          >
            Favorite
          </button>
        ) : (
          <button
            className="unfavorite"
            onClick={() => props.removeFromFavorites(name)}
          >
            ✓
          </button>
        )}
      </article>
    );
  }
}

Card.propTypes = {
  format: PropTypes.string.isRequired,
  element: PropTypes.object.isRequired,
  addToFavorites: PropTypes.func.isRequired,
  removeFromFavorites: PropTypes.func.isRequired
};
