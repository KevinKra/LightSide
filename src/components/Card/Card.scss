.Card {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  border: 1px solid black;
  min-width: 300px;
  max-width: 300px;
  z-index: 1;
  margin: 0 0.5rem;
  height: 80vh;
  color: white;
  button {
    border: none;
    color: white;
    border-top: 1px solid black;
    background-color: rgba(0, 0, 0, 0.326);
    padding: 0.8rem 0;
    transition: 0.09s background-color ease-in;
    cursor: pointer;
    &:hover {
      background-color: rgba(0, 0, 0, 0.726);
    }
    &:focus {
      outline: none;
    }
  }
}

.card-bg {
  position: absolute;
  z-index: -1;
  height: 100%;
  width: 100%;
  background: linear-gradient(
    180deg,
    rgba(0, 0, 0, 1) 0%,
    rgba(0, 0, 0, 1) 40%,
    rgba(0, 0, 0, 0.542454481792717) 83%,
    rgba(0, 0, 0, 0) 100%
  );
}
