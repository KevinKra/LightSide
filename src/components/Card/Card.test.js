import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Card from "../Card/Card";

const mockPerson = {
  name: "Joe",
  gender: "Male",
  hair_color: "Green",
  eye_color: "Blue",
  birth_year: 1995,
  mass: 120,
  height: 72
};

export const mockVehicle = {
  name: "Speeder",
  model: "Super",
  vehicle_class: "X12",
  manufacturer: "Toyota",
  const_in_credits: 15000,
  max_atmosphere_speed: 50,
  crew: 15,
  passengers: 30,
  cargo_capacity: 5
};

const mockPlanet = {
  name: "Naboo",
  climate: "Moderate",
  terrain: "Grasslands",
  population: 5000000,
  rotation_period: 24,
  orbital_period: 375,
  diameter: 15,
  favorited: false
};

describe("<Card />", () => {
  it("should render in the person format", () => {
    const wrapper = shallow(<Card format="person" element={mockPerson} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should render in the vehicle format", () => {
    const wrapper = shallow(<Card format="vehicle" element={mockVehicle} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should render in the planet format", () => {
    const wrapper = shallow(<Card format="planet" element={mockPlanet} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should render text 'Favorite' if button's element.favorite is true", () => {
    const mockData = [mockPerson, mockPlanet, mockVehicle];
    const wrapperPerson = shallow(
      <Card format="person" element={mockData[0]} />
    );
    const wrapperPlanet = shallow(
      <Card format="planet" element={mockData[1]} />
    );
    const wrapperVehicle = shallow(
      <Card format="vehicle" element={mockData[2]} />
    );
    const favoritedCards = [wrapperPerson, wrapperPlanet, wrapperVehicle];
    favoritedCards.forEach(card => {
      const button = card.find(".favorite");
      expect(button.text().includes("Favorite")).toBe(true);
    });
  });

  it("shoulder render text '✓' if button's element.favroite is false", () => {
    const mockData = [mockPerson, mockPlanet, mockVehicle];
    mockData.forEach(mock => (mock.favorited = true));
    const wrapperPerson = shallow(
      <Card format="person" element={mockData[0]} />
    );
    const wrapperPlanet = shallow(
      <Card format="planet" element={mockData[1]} />
    );
    const wrapperVehicle = shallow(
      <Card format="vehicle" element={mockData[2]} />
    );

    const favoritedCards = [wrapperPerson, wrapperPlanet, wrapperVehicle];
    favoritedCards.forEach(card => {
      const button = card.find(".unfavorite");
      expect(button.text().includes("✓")).toBe(true);
    });
  });
});
