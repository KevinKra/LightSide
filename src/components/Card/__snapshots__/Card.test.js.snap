// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render in the person format 1`] = `
<article
  className="Card"
>
  <section
    className="card-primary"
  >
    <h1>
      Joe
    </h1>
    <p>
      gender: 
      Male
    </p>
    <p>
      hair: 
      Green
    </p>
    <p>
      eye color: 
      Blue
    </p>
    <p>
      birth year: 
      1995
    </p>
    <p>
      mass: 
      120
    </p>
    <p>
      height: 
      72
    </p>
  </section>
  <div
    className="card-bg person-bg"
  />
  <button
    className="favorite"
    onClick={[Function]}
  >
    Favorite
  </button>
</article>
`;

exports[`<Card /> should render in the planet format 1`] = `
<article
  className="Card"
>
  <section
    className="card-primary"
  >
    <h1>
      Naboo
    </h1>
    <p>
      climate: 
      Moderate
    </p>
    <p>
      terrain: 
      Grasslands
    </p>
    <p>
      population: 
      5000000
    </p>
    <p>
      day length: 
      24
       hours
    </p>
    <p>
      year length: 
      375
       days
    </p>
    <p>
      diameter: 
      15
    </p>
  </section>
  <div
    className="card-bg planet-bg"
  />
  <button
    className="favorite"
    onClick={[Function]}
  >
    Favorite
  </button>
</article>
`;

exports[`<Card /> should render in the vehicle format 1`] = `
<article
  className="Card"
>
  <p
    className="card-primary"
  >
    <h1>
      Speeder
    </h1>
    <p>
      model: 
      Super
    </p>
    <p>
      class: 
      X12
    </p>
    <p>
      manufacturer: 
      Toyota
    </p>
    <p>
      cost: 
    </p>
    <p>
      maximum speed: 
    </p>
    <p>
      crew capacity: 
      15
    </p>
    <p>
      passenger capacity: 
      30
    </p>
    <p>
      cargo capacity: 
      5
    </p>
  </p>
  <div
    className="card-bg vehicle-bg"
  />
  <button
    className="favorite"
    onClick={[Function]}
  >
    Favorite
  </button>
</article>
`;

exports[`<Card /> should render text '✓' if button's element.favorite is false 1`] = `
<article
  className="Card"
>
  <section
    className="card-primary"
  >
    <h1>
      Joe
    </h1>
    <p>
      gender: 
      Male
    </p>
    <p>
      hair: 
      Green
    </p>
    <p>
      eye color: 
      Blue
    </p>
    <p>
      birth year: 
      1995
    </p>
    <p>
      mass: 
      120
    </p>
    <p>
      height: 
      72
    </p>
  </section>
  <div
    className="card-bg person-bg"
  />
  <button
    className="unfavorite"
    onClick={[Function]}
  >
    ✓
  </button>
</article>
`;

exports[`<Card /> should render text '✓' if button's element.favorite is false 2`] = `
<article
  className="Card"
>
  <section
    className="card-primary"
  >
    <h1>
      Naboo
    </h1>
    <p>
      climate: 
      Moderate
    </p>
    <p>
      terrain: 
      Grasslands
    </p>
    <p>
      population: 
      5000000
    </p>
    <p>
      day length: 
      24
       hours
    </p>
    <p>
      year length: 
      375
       days
    </p>
    <p>
      diameter: 
      15
    </p>
  </section>
  <div
    className="card-bg planet-bg"
  />
  <button
    className="unfavorite"
    onClick={[Function]}
  >
    ✓
  </button>
</article>
`;

exports[`<Card /> should render text '✓' if button's element.favorite is false 3`] = `
<article
  className="Card"
>
  <p
    className="card-primary"
  >
    <h1>
      Speeder
    </h1>
    <p>
      model: 
      Super
    </p>
    <p>
      class: 
      X12
    </p>
    <p>
      manufacturer: 
      Toyota
    </p>
    <p>
      cost: 
    </p>
    <p>
      maximum speed: 
    </p>
    <p>
      crew capacity: 
      15
    </p>
    <p>
      passenger capacity: 
      30
    </p>
    <p>
      cargo capacity: 
      5
    </p>
  </p>
  <div
    className="card-bg vehicle-bg"
  />
  <button
    className="unfavorite"
    onClick={[Function]}
  >
    ✓
  </button>
</article>
`;
